feat: add CLI integration tests using assert_cmd

mFragaBA opened the pull request:
closes #335

igamigo commented on tests/integration/cli_tests.rs:
```diff
@@ -0,0 +1,372 @@
+#[test]
+fn test_init_without_params() {
+    // For now sleep to ensure node's up
+    std::thread::sleep(std::time::Duration::new(30, 0));
```
I think we should try to avoid sleeping as much as possible. Could we poll with `miden sync` until it doesn't fail? In this specific case however, if we are just testing the `init` command it's probably not necessary to wait at all since it's independent of the node (and we also want to test that these commands actually work without the node being up as well).

mFragaBA replied:
since we do the sync afterwards to check the init was done successfully,  I'd still need the node to be up, but what I ended up doing is modifying the `sync_state` function from this module so it syncs until the command finally succeeds. If it never succeeds nextest will catch because of the test timeout and it'll fail.

mFragaBA commented on tests/integration/cli_tests.rs:
```diff
@@ -0,0 +1,372 @@
+    };
+
+    // Ensure they've been importing by showing them
+    // TODO: Once show is fixed for faucet account do the full iteration without skipping the
```
There is no easy way to detect that (By easy I mean without altering the Client/Store API), and considering in base on next it's already possible to print AccountCode without import info I don't think it's worth it. Maybe a simplification would be to hide account code for faucets, but it's still not a great solution.

mFragaBA commented on tests/integration/cli_tests.rs:
```diff
@@ -0,0 +1,372 @@
+        &second_basic_account_id,
+        "--force",
+        &note_to_consume_id,
+    ]);
```
done! I made it so  and consume always use partial note ids. For mint I always use the full provided account id (but call the function with the shortened ids whenever possible) and added a test that uses an untracked account with the full ID

mFragaBA commented on tests/integration/cli_tests.rs:
```diff
@@ -0,0 +1,372 @@
+        "-n",
+        "private",
+        "--force",
+    ]);
```
done!

mFragaBA commented on tests/integration/cli_tests.rs:
```diff
@@ -0,0 +1,372 @@
+    std::fs::create_dir(temp_dir.clone()).unwrap();
+
+    let mut init_cmd = Command::cargo_bin("miden").unwrap();
+    init_cmd.args(["init", "--rpc", "localhost", "--store-path", store_path.to_str().unwrap()]);
```
nice suggestion! Done!

mFragaBA commented on tests/integration/cli_tests.rs:
```diff
@@ -0,0 +1,372 @@
+    temp_dir.push(format!("{}", uuid::Uuid::new_v4()));
+    std::fs::create_dir(temp_dir.clone()).unwrap();
+
+    let mut init_cmd = Command::cargo_bin("miden").unwrap();
+    init_cmd.args(["init"]);
+    init_cmd.current_dir(&temp_dir).assert().success();
```
didn't think of that but i like the idea

mFragaBA commented on src/main.rs:
```diff
@@ -4,13 +4,13 @@ mod cli;
-    }
+    cli.execute().await?;
+
+    Ok(())
```
done!

mFragaBA commented:
>     * In cases where the client sync got stuck, the node was being spammed with requests.

Nice catch! I added a sleep in between syncs

>     * The `test_import_pending_notes` failed despite not being a `cli_test`.

we debugged this on a huddle and it was a bug when traversing the mmr indices on `get_and_store_authenticated_block`. Fixed it on [24fe59e](https://github.com/0xPolygonMiden/miden-client/pull/353/commits/24fe59e4079384bf146b215b64e2a5afd4a46239) with @igamigo suggestion.

>     * Is there a way to run the genesis account test without restarting the node each time?

We could change the genesis accounts so they're public accounts, but we'd need to use the node on `next` because [#365](https://github.com/0xPolygonMiden/miden-node/pull/365) got merged to next but isn't on next yet.

>     * Is there a way to run the tests with a node that isn't in the miden-client directory? (e.g. the local docker node)

you could technically run them with a dockerized node by copying the genesis account files out and running:

```shell
mkdir -p ./miden-node/accounts
cp <path_to_genesis_account_files_extracted_from_node> ./miden-node/accounts/
cargo make integration-tests
```

that's kind of anoying though, and still wouldn't allow you to run integration tests against a remote node. Perhaps it'd be better to set the genesis test as ignored and have a Makefile command to run integration tests with the genesis test included and without. What do you think? Are there any other possibilities?

mFragaBA commented:
@igamigo @tomyrd  I ended up going with ignoring the test by default and enabling it on the `integration-test-full` target. `integration-test-full` is the one ran on the CI and we leave the `integration-test` to run without the ignored test

igamigo commented on Makefile.toml:
```diff
@@ -49,10 +49,13 @@ description = "Run the client testing suite"
-description = "Run the integration test binary. Requires a node to connect to."
-command = "cargo"
-args = ["nextest", "run", "--profile", "ci-default", "--release", "--test=integration", "--features", "${FEATURES_INTEGRATION_TESTING}"]
+[tasks.integration-test-full]
+description = "Run the integration test binary with ignored tests included"
+script_runner = "bash"
+script = [
+  'cargo nextest run --profile ci-default --release --test=integration --features $FEATURES_INTEGRATION_TESTING',
+  'cargo nextest run --profile ci-default --release --test=integration --features $FEATURES_INTEGRATION_TESTING --run-ignored ignored-only -- test_import_genesis_accounts_can_be_used_for_transactions'
+]
```
Instead of doing this, could we change the nextest profile instead to include ignored tests from the binary?

https://nexte.st/book/other-options?highlight=ignored#runner-options

bobbinth replied:
If we can control which tests to run via nextest profiles, may be good to apply the same approach in `miden-base` too (there we run slow proving tests and regular tests in different profiles).

mFragaBA replied:
nextest lets you control whether to run all ignored, non ignored or all tests but doesn't let you specify which. That's why I had to use a second command with the `--run-ignored ignored-only` (this is needed or the ignored test won't run) and the `-- test_import_genesis_accounts_can_be_used_for_transactions` (this to select the test by name)

bobbinth replied:
But can we specify in a profile (not via command line args) to run only ignored tests?

mFragaBA replied:
I looked and the profiles don't have a run-ignored setting. Also checked for a workaround but profiles don't seem to let you define whether something gets run or not. Sad because you can do overrides with

```toml
[proifle.default]
filter = test(not =test_import_genesis_accounts_can_be_used_for_transactions)
```

mFragaBA replied:
I think we can merge as is and look for alternatives if it becomes difficult to mantain

bobbinth replied:
> I think we can merge as is and look for alternatives if it becomes difficult to mantain

I'm fine with this.

> ```
> [proifle.default]
> filter = test(not =test_import_genesis_accounts_can_be_used_for_transactions)
> ```

To clarify: is something like this currently possible or not possible?

mFragaBA replied:
you can define it but I think that won't filter the tests that are ran, just filters who gets that profile. For example:

```yaml
[proifle.default]
filter = test(not =test_import_genesis_accounts_can_be_used_for_transactions)
slow-timeout = { period = "3m", terminate-after = 4}
```

That just means that when using the default profile, you apply that slow-timeout setting to all tests that are not `test_import_genesis_accounts_can_be_used_for_transactions`. But it'll still run `test_import_genesis_accounts_can_be_used_for_transactions`, just not with that setting

mFragaBA replied:
It'd be nice if they had a setting to skip a test run

igamigo reviewed (approved):
LGTM!
